libsoftwarecontainer/src/gateway/dbus/dbusgatewayinstance.cpp

```diff
@@ -81,62 +91,75 @@ bool DBusGatewayInstance::readConfigElement(const json_t *element)
+    // updated.
+    if (!m_activatedOnce) {
+        // set DBUS_{SESSION,SYSTEM}_BUS_ADDRESS env variable
+        std::string variable = std::string("DBUS_")
+                             + (m_type == SessionProxy ? "SESSION" : "SYSTEM")
+                             + std::string("_BUS_ADDRESS");
+        std::string value = "unix:path=" + buildPath("/gateways", socketName());
+        getContainer()->setEnvironmentVariable(variable, value);
+
+        std::vector<std::string> commandVec = {"dbus-proxy",
+                                               m_socket,
+                                               m_type == SessionProxy ? "session" : "system"};
+
+        // Give the dbus-proxy access to the real dbus bus address.
+        std::vector<std::string> envVec;
+
+        bool hasEnvVar = false;
+        std::string envValue = Glib::getenv(variable, hasEnvVar);
+
+        if (!hasEnvVar) {
+            if (m_type == SessionProxy) {
+                log_error() << "Using DBus gateway in session mode"
+                            << " and no " + variable + " set in host environment, dbus-proxy won't work";
+                return false;
+            } else {
+                log_warn() << "Using DBus gateway in system mode"
+                           << " and no " + variable + " set in host environment, this could be a problem";
+            }
         } else {
-            log_warn() << "Using DBus gateway in system mode"
-                       << " and no " + variable + " set in host environment, this could be a problem";
+            envVec.push_back(variable + "=" + envValue);
         }
-    } else {
-        envVec.push_back(variable + "=" + envValue);
-    }
 
-    if (!startDBusProxy(commandVec, envVec)) {
-        return false;
+        if (!startDBusProxy(commandVec, envVec)) {
+            return false;
+        }
     }
 
-    // Dump to string and clean up
+    // Dump to string. Compact format is currently needed as dbus-proxy relies on
+    // the config string to be without newlines anywhere in the middle of it all
     char *config_c = json_dumps(m_entireConfig, JSON_COMPACT);
     std::string config = std::string(config_c);
+
     free(config_c);
 
     return testDBusConnection(config);
 }
 
-bool DBusGatewayInstance::testDBusConnection(const std::string &config)
+bool DBusGatewayInstance::testDBusConnection(const std::string &cfg)
 {
-    log_debug() << "Sending config " << config;
-    unsigned int count = sizeof(char) * config.length();
-    ssize_t written = write(m_infp, config.c_str(), count);
+    std::stringstream ss;
+    ss << cfg << std::string("\n");
+    std::string config = ss.str();
+
+    unsigned char configLength = config.length();
+    ssize_t count = (sizeof(char) * configLength);
+    log_debug() << "Expected config byte length " << count;
+
+    log_debug() << "Config: " << config;
+
+    ssize_t configWrite = write(m_proxyStdin, config.c_str(), count);
```

Comment:
maybe it would be more concise like
std::string config{cfg + std::string("\n")}
ssize_t configWrite = write(m_proxyStdin, config.c_str(), config.length());